[SPARK-35639][SQL] Add metrics about coalesced partitions to AQEShuffleRead in AQE

ekoifman opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://spark.apache.org/contributing.html
  2. Ensure you have added or run the appropriate tests for your PR: https://spark.apache.org/developer-tools.html
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][SPARK-XXXX] Your PR title ...'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
  7. If you want to add a new configuration, please read the guideline first for naming configurations in
     'core/src/main/scala/org/apache/spark/internal/config/ConfigEntry.scala'.
  8. If you want to add or modify an error message, please read the guideline first:
     https://spark.apache.org/error-message-guidelines.html
-->

### What changes were proposed in this pull request?
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->
AQEShuffleReadExec already reports "number of skewed partitions" and "number of skewed partition splits".
It would be useful to also report "number of coalesced partitions" and for ShuffleExchange to report "number of partitions"
This way it's clear what happened on the map side and on the reduce side.

![Metrics](https://user-images.githubusercontent.com/4297661/126729820-cf01b3fa-7bc4-44a5-8098-91689766a68a.png)



### Why are the changes needed?
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->
Improves usability

### Does this PR introduce _any_ user-facing change?
<!--
Note that it means *any* user-facing change including all aspects such as the documentation fix.
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released Spark versions or within the unreleased branches such as master.
If no, write 'No'.
-->
Yes, it now provides more information about `AQEShuffleReadExec` operator behavior in the metrics system.

### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
-->
Existing tests

ekoifman committed:
[SPARK-35639][SQL] Add metrics about coalesced partitions to CustomShuffleReader in AQE
[SPARK-35639][SQL] added numPartitions metric to ShuffleExchangeExec; addressed review comments
[SPARK-35639][SQL] rebased on master

AmplabJenkins commented:
Can one of the admins verify this patch?

ekoifman commented:
@cloud-fan could you please take a look

ulysses-you commented on sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/CustomShuffleReaderExec.scala:
```diff
@@ -76,19 +76,76 @@ case class CustomShuffleReaderExec private(
+  }
 
-  def hasCoalescedPartition: Boolean =
-    partitionSpecs.exists(_.isInstanceOf[CoalescedPartitionSpec])
```
this change worth a new PR, can you create a new one for it ? thanks

ekoifman replied:
I don't understand this comment.  This change is a critical part of this PR

ulysses-you replied:
This PR include two parts, isn't it ?
*  the result of if `CustomShuffleReaderExec` is coalesced changed (it's an exists issue)
*  extra metrics with `CustomShuffleReaderExec` (it's an improvement)

ekoifman replied:
this was done via https://github.com/apache/spark/pull/32872

cloud-fan commented:
ok to test

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/CustomShuffleReaderExec.scala:
```diff
@@ -76,19 +76,76 @@ case class CustomShuffleReaderExec private(
       "local"
     } else if (hasCoalescedPartition && hasSkewedPartition) {
-      "coalesced and skewed"
+      s"$coalescedDetail and $skewedDetail"
```
It makes sense to add more metrics but it doesn't make sense to include metrics in the plan node string.

cloud-fan replied:
AFAIK plan node string never contains metrics before this PR.

ekoifman replied:
The reason I put this additional info here is that it's easy to see what AQE did in large (100s of operators) plans in the logged version of the plan.  It puts that info in the right context - the 1st image in the PR description shows what it looks like.  
If the additional info was not exposed as SQL metrics but simply computed just to put in the node description - would it still be wrong to put it there?

cloud-fan replied:
No other plan node does this: metrics are only available through UI or programing APIs (e.g. `QueryExecutionListener`).

I can't find a strong reason to start doing it.

ekoifman replied:
remved `stringArgs` changes

SparkQA commented:
**[Test build #139613 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139613/testReport)** for PR 32776 at commit [`aee1392`](https://github.com/apache/spark/commit/aee1392720815f332e8fb993b4672bb03fe4ccb1).

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/CustomShuffleReaderExec.scala:
```diff
@@ -171,6 +238,19 @@ case class CustomShuffleReaderExec private(
+          if (hasCoalescedPartition) {
+            Map("numCoalescedPartitions" ->
+              SQLMetrics.createMetric(sparkContext, "number of coalesced partitions"),
+              "numPartitionsToCoalesce" ->
```
There is always a shuffle node below `CustomShuffleReader`. I think it makes more sense to let the shuffle node to report the metrics of the number of partitions/reducers.

ekoifman replied:
I added the coalesce metrics in the same way skew-related metrics were done.
Are you suggesting that all of them move to the `ShuffleExchangeExec` below?

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/CustomShuffleReaderExec.scala:
```diff
@@ -76,19 +76,76 @@ case class CustomShuffleReaderExec private(
+   * MapOutputStatistics 0,2,0,2,7
+   * Results in CoalescedPartitionSpec(1,4) & CoalescedPartitionSpec(4,5)
+   * So pre-shuffle partition 2 is included
+   * We could figure out dropped partitions but doesn't seem that useful.
```
I don't think it's useful to report this metrics.

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/44140/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/44140/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/44140/

SparkQA commented:
**[Test build #139613 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/139613/testReport)** for PR 32776 at commit [`aee1392`](https://github.com/apache/spark/commit/aee1392720815f332e8fb993b4672bb03fe4ccb1).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/139613/

SparkQA commented:
**[Test build #141100 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/141100/testReport)** for PR 32776 at commit [`5360f51`](https://github.com/apache/spark/commit/5360f5127a69fd1468f3bb17f3b822e1edeffdad).

ekoifman commented:
@cloud-fan 
I rebased after https://github.com/apache/spark/pull/32872 and https://github.com/apache/spark/pull/33342
and removed the changes to `CustomShuffleReaderExec.stringArgs` that you objected to.

So now this just adds 2 new metrics `numCoalescedPartitions` and `numPartitionsToCoalesce` to compliment the existing `numSkewedSplits` metrics.

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/45613/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/45613/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/45613/

SparkQA commented:
**[Test build #141100 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/141100/testReport)** for PR 32776 at commit [`5360f51`](https://github.com/apache/spark/commit/5360f5127a69fd1468f3bb17f3b822e1edeffdad).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/141100/

cloud-fan commented:
Since skewed partitions and coalesced partitions can appear together in one custom shuffle reader, I think it's reasonable to add the `numCoalescedPartitions` metric.

`numPartitionsToCoalesce` is a bit overkill to me. If we can know the original number of partitions of the shuffle (ShuffleExchangeExec can report it), we can get a sense of the custom shuffle reader by looking at num skewed partitions/splits and num coalesced partitions.

ekoifman commented:
That makes sense.

SparkQA commented:
**[Test build #141526 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/141526/testReport)** for PR 32776 at commit [`83a4082`](https://github.com/apache/spark/commit/83a40829101facba700463dfcb1d2c803d7fc157).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/46044/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/46044/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/46044/

SparkQA commented:
**[Test build #141526 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/141526/testReport)** for PR 32776 at commit [`83a4082`](https://github.com/apache/spark/commit/83a40829101facba700463dfcb1d2c803d7fc157).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/141526/

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/CustomShuffleReaderExec.scala:
```diff
@@ -182,6 +193,17 @@ case class CustomShuffleReaderExec private(
           Map.empty
         }
+      } ++ {
+        if (isLocalReader) {
```
do we need this if condition here?

ekoifman replied:
I think so.  For local reader all `CustomShuffleReaderExec.partitionSpecs` should be `PartialMapperPartitionSpec`
so `hasCoalescedPartitions` is false, so `numCoalescedPartitions` metric would always be `0`.  It doesn't seem to add value to report it.

ekoifman commented:
rebased on master after SPARK-36217

SparkQA commented:
**[Test build #141655 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/141655/testReport)** for PR 32776 at commit [`18f865b`](https://github.com/apache/spark/commit/18f865bdfcf67b7f8c330257aae214f3773857fc).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/46171/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/46171/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/46171/

SparkQA commented:
**[Test build #141655 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/141655/testReport)** for PR 32776 at commit [`18f865b`](https://github.com/apache/spark/commit/18f865bdfcf67b7f8c330257aae214f3773857fc).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/141655/

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AQEShuffleReadExec.scala:
```diff
@@ -183,6 +194,17 @@ case class AQEShuffleReadExec private(
+        if (isLocalRead) {
+          Map.empty
+        } else {
+          if (hasCoalescedPartition) {
```
reply to https://github.com/apache/spark/pull/32776#discussion_r676750494

If this is a local reader, then `hasCoalescedPartition` can never be true, right? I still think the `if (isLocalRead)` above is not needed.

ekoifman replied:
yes, you're right

ekoifman replied:
updated

ekoifman committed:
[SPARK-35639][SQL] removed unnesessary 'if'

SparkQA commented:
**[Test build #141713 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/141713/testReport)** for PR 32776 at commit [`7f51402`](https://github.com/apache/spark/commit/7f5140228c5f4f5876dd6653d90472fc92bdb5f4).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/46226/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/46226/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/46226/

cloud-fan commented on sql/core/src/test/scala/org/apache/spark/sql/ExplainSuite.scala:
```diff
@@ -572,6 +572,7 @@ class ExplainSuiteAE extends ExplainSuiteHelper with EnableAdaptiveExecutionSuit
       """
         |(11) AQEShuffleRead
         |Input [5]: [k#x, count#xL, sum#xL, sum#x, count#xL]
+        |Arguments: coalesced
```
just for curiosity: which code change leads to this?

ekoifman replied:
There isn't one in this PR.  This line was part of the actual output before. The test checks if the actual plan contains
```
      """
        |(11) AQEShuffleRead
        |Input [5]: [k#x, count#xL, sum#xL, sum#x, count#xL]
        |Arguments: coalesced
        |""".stripMargin,
```
which is true in both versions.

SparkQA commented:
**[Test build #141713 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/141713/testReport)** for PR 32776 at commit [`7f51402`](https://github.com/apache/spark/commit/7f5140228c5f4f5876dd6653d90472fc92bdb5f4).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/141713/

cloud-fan commented:
thanks, merging to master/3.2!
